File: zh-hant/codes/python/chapter_backtracking/preorder_traversal_iii_template.py
"""
File: preorder_traversal_iii_template.py
Created Time: 2023-04-15
Author: krahets (krahets@163.com)
"""

import sys
from pathlib import Path

sys.path.append(str(Path(__file__).parent.parent))
from modules import TreeNode, print_tree, list_to_tree


def is_solution(state: list[TreeNode]) -> bool:
    """判斷當前狀態是否為解"""
    return state and state[-1].val == 7


def record_solution(state: list[TreeNode], res: list[list[TreeNode]]):
    """記錄解"""
    res.append(list(state))


def is_valid(state: list[TreeNode], choice: TreeNode) -> bool:
    """判斷在當前狀態下，該選擇是否合法"""
    return choice is not None and choice.val != 3


def make_choice(state: list[TreeNode], choice: TreeNode):
    """更新狀態"""
    state.append(choice)


def undo_choice(state: list[TreeNode], choice: TreeNode):
    """恢復狀態"""
    state.pop()


def backtrack(
    state: list[TreeNode], choices: list[TreeNode], res: list[list[TreeNode]]
):
    """回溯演算法：例題三"""
    # 檢查是否為解
    if is_solution(state):
        # 記錄解
        record_solution(state, res)
    # 走訪所有選擇
    for choice in choices:
        # 剪枝：檢查選擇是否合法
        if is_valid(state, choice):
            # 嘗試：做出選擇，更新狀態
            make_choice(state, choice)
            # 進行下一輪選擇
            backtrack(state, [choice.left, choice.right], res)
            # 回退：撤銷選擇，恢復到之前的狀態
            undo_choice(state, choice)


"""Driver Code"""
if __name__ == "__main__":
    root = list_to_tree([1, 7, 3, 4, 5, 6, 7])
    print("\n初始化二元樹")
    print_tree(root)

    # 回溯演算法
    res = []
    backtrack(state=[], choices=[root], res=res)

    print("\n輸出所有根節點到節點 7 的路徑，要求路徑中不包含值為 3 的節點")
    for path in res:
        print([node.val for node in path])

File: zh-hant/codes/rust/chapter_backtracking/preorder_traversal_iii_template.rs
/*
 * File: preorder_traversal_iii_template.rs
 * Created Time: 2023-07-15
 * Author: codingonion (coderonion@gmail.com)
 */

use hello_algo_rust::include::{print_util, vec_to_tree, TreeNode};
use std::{cell::RefCell, rc::Rc};

/* 判斷當前狀態是否為解 */
fn is_solution(state: &mut Vec<Rc<RefCell<TreeNode>>>) -> bool {
    return !state.is_empty() && state.last().unwrap().borrow().val == 7;
}

/* 記錄解 */
fn record_solution(
    state: &mut Vec<Rc<RefCell<TreeNode>>>,
    res: &mut Vec<Vec<Rc<RefCell<TreeNode>>>>,
) {
    res.push(state.clone());
}

/* 判斷在當前狀態下，該選擇是否合法 */
fn is_valid(_: &mut Vec<Rc<RefCell<TreeNode>>>, choice: Option<&Rc<RefCell<TreeNode>>>) -> bool {
    return choice.is_some() && choice.unwrap().borrow().val != 3;
}

/* 更新狀態 */
fn make_choice(state: &mut Vec<Rc<RefCell<TreeNode>>>, choice: Rc<RefCell<TreeNode>>) {
    state.push(choice);
}

/* 恢復狀態 */
fn undo_choice(state: &mut Vec<Rc<RefCell<TreeNode>>>, _: Rc<RefCell<TreeNode>>) {
    state.pop();
}

/* 回溯演算法：例題三 */
fn backtrack(
    state: &mut Vec<Rc<RefCell<TreeNode>>>,
    choices: &Vec<Option<&Rc<RefCell<TreeNode>>>>,
    res: &mut Vec<Vec<Rc<RefCell<TreeNode>>>>,
) {
    // 檢查是否為解
    if is_solution(state) {
        // 記錄解
        record_solution(state, res);
    }
    // 走訪所有選擇
    for &choice in choices.iter() {
        // 剪枝：檢查選擇是否合法
        if is_valid(state, choice) {
            // 嘗試：做出選擇，更新狀態
            make_choice(state, choice.unwrap().clone());
            // 進行下一輪選擇
            backtrack(
                state,
                &vec![
                    choice.unwrap().borrow().left.as_ref(),
                    choice.unwrap().borrow().right.as_ref(),
                ],
                res,
            );
            // 回退：撤銷選擇，恢復到之前的狀態
            undo_choice(state, choice.unwrap().clone());
        }
    }
}

/* Driver Code */
pub fn main() {
    let root = vec_to_tree([1, 7, 3, 4, 5, 6, 7].map(|x| Some(x)).to_vec());
    println!("初始化二元樹");
    print_util::print_tree(root.as_ref().unwrap());

    // 回溯演算法
    let mut res = Vec::new();
    backtrack(&mut Vec::new(), &mut vec![root.as_ref()], &mut res);

    println!("\n輸出所有根節點到節點 7 的路徑，要求路徑中不包含值為 3 的節點");
    for path in res {
        let mut vals = Vec::new();
        for node in path {
            vals.push(node.borrow().val)
        }
        println!("{:?}", vals);
    }
}
